torch/csrc/jit/codegen/cuda/runtime/block_reduction.cu

```diff
@@ -89,8 +89,8 @@ __device__ void blockReduce(
     }
   }
   __syncthreads();
-  // for (int factor = np2/2; factor > contig_threads / 2; factor>>=1) {
-  for (int factor = np2 / 2; factor > 0; factor >>= 1) {
+  // loop peel the final iteration to save one syncthread for the end
```

Comment:
(This is just my general comment, and not requesting any change.) I'd prefer to keep a simpler implementation of these collective routines for now. For optimizing performance, we should probably use hardware-supported collectives anyway. Reducing just one time of syncthreads seems premature optimization to me.